### ISSUE TITLE
Implement extract for a type that implements From<Vec<u8>>?

### ISSUE THREAD

**Issue by nerdrew** [carllerche/tower-web#117]
I have a type that implements `From<Vec<u8>>`. Do I need to write a full `impl Extract` and `impl ExtractFuture` for my type? It looks like I can mostly just copy the `bytes` extract but that seems a bit suboptimal.

If the `extract::bytes::*` was public, I think I can just do:

```rust
impl<B: BufStream> Extract<B> for MyStruct {
    type Future = ExtractBytes<Self, B>;

    fn extract(ctx: &Context) -> Self::Future {
        unimplemented!();
    }

    fn extract_body(ctx: &Context, body: B) -> Self::Future {
        let state = State::Body(body.collect());
        ExtractBytes { state }
    }

    fn requires_body(callsite: &CallSite) -> bool {
        true
    }
}
```

**Comment by carllerche** [comment]
Yeah, those types probably could be public. I also think we could support this case w/ `#[derive(Extract)]`:

```rust
#[derive(Extract)]
#[web(from_bytes)]
struct MyType { ... }
```

In this case, you would implement `From<Bytes>`, which is mostly equivalent. Though, if you really need to convert from a vec, that could possibly be supported too. Could you share more details about the type that can be converted from `Vec<u8>`?

**Comment by nerdrew** [comment]
`MyType` is a generated protobuf message struct.

**Comment by qmx** [comment]
That would be useful to me too, I'm doing the same thing but with `Vec<u8>` being a snappy-compressed protobuf message

**Comment by nerdrew** [comment]
I'm playing around with [quick-protobuf](https://github.com/tafia/quick-protobuf) and [rental](https://github.com/jpernst/rental) together. I can't tell if it is a winning combination or a terrible idea.

My struct looks like this:

```rust
rental! {
    mod rentals {
        use super::*;

        #[rental(covariant, debug)]
        pub struct MyProtoRental{
            buf: Vec<u8>,
            proto: MyProtoMessage<'buf>,
        }
    }
}
```

Unfortunately, the way rental works I can't add other attributes on my struct (I don't know if that is just lack of a feature or a permanent issue).

If I need to implement `From<Bytes>` for `MyProtoRental`, is it safe to implement `StableDeref` for `Bytes`? It looks like it probably isn't since short messages aren't heap allocated. Is it possible to guarantee all `Bytes` that get used in my `From` are on the heap (and therefore safe to mark `StableDeref`)?

I was playing with warp and I implemented `StableDeref` + `Deref` for a newtype around [FullBody](https://docs.rs/warp/0.1.3/warp/filters/body/struct.FullBody.html). I'm not sure if that is safe.

**Comment by carllerche** [comment]
I haven't forgotten this, I think protobuf is a good usecase.

I don't have an immediate great solution. Supporting `From<Vec<u8>>` would work, but that wouldn't take advantage of `Bytes` and zero copy.

I want to spend some more time thinking about it.

In the mean time, taking `body: Vec<u8>` works... not great but works.

**Comment by carllerche** [comment]
@qmx @nerdrew Instead of implementing `Extract` for `From<Vec<u8>>`, I'm thinking it would be for `FromBufStream` (being proposed here: https://github.com/tokio-rs/tokio/pull/611/files#diff-026fa19f4a16876a112c89b16c89bccaR15).

Pro: It is more flexible than `T: From<Vec<u8>>`.

Con: It would require types to implement `FromBufStream`.

To use with protobuf, you would need the lib in question to provide an implementation of `FromBufStream` for their types. This could be provided via a feature flag.

The work around is always a wrapper type.

**Comment by nerdrew** [comment]
I was using quick-protobuf b/c I like that it doesn't allocate `String`s and just uses `Cow<str>` when decoding string fields. Though if [this prost issue](https://github.com/danburkert/prost/issues/31) gets done, I wonder if there is any performance benefit to `Cow` vs using bytes + your string crate.

**Comment by nerdrew** [comment]
I haven't tried `bytes` out yet. When decoding a protobuf with `FromBufStream`, would you need to incrementally decode the proto as `extend` is called? Also, related to my previous [post](https://github.com/carllerche/tower-web/issues/117#issuecomment-429371414), if the proto struct uses `bytes::Bytes` (not sure if the fields would be `Bytes` or `BytesMut`) and `string::String` fields, could the `FromBufStream` be zero copy when decoding those fields?

**Comment by carllerche** [comment]
`std::String` fields cannot be zero copy when coming from `Bytes`. However, you can also use the `string` crate (http://crates.io/crates/string) which does support zero copy. In that case, you would have `String<Bytes>`.
